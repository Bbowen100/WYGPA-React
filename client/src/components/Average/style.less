.body {
  text-align: center;
}

.average {
  font-size: 50;
}
