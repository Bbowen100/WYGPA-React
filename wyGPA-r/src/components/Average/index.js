import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {updateAverage} from '../../actions';

class Average extends Component {
  constructor(props){
    super(props);
    this.state = {
      average: null,
    };
  }

  shouldComponentUpdate(nextProps, nextState){
    return !(nextState.average === this.state.average);
  }

  calcAvg(source){
    let avg = 0;
    if(source.items && source.items.length > 0){
      let totMark = source.items.map((ele)=>ele.mark*ele.weight).reduce((acc, val)=>acc+val);
      let totWeight = source.items.map(ele=>ele.weight).reduce((acc, val)=>acc+val);
      avg = Number((totMark / totWeight).toFixed(2));
    }
    if(this.state.average !== avg) {
      this.setState({average: avg}, ()=>{
        this.props.dispatch(updateAverage(avg));
      });
    }
  }

  componentDidMount(){
    this.calcAvg(this.props);
  }

  componentWillReceiveProps(newProps){
    this.calcAvg(newProps);
  }

  render(){
    return(
      <div>
        <label>Your average is: {this.state.average} </label>
      </div>
    )
  }
}

Average.PropTypes = {
  items: PropTypes.arrayOf(PropTypes.objectOf(PropTypes.shape({
    name: PropTypes.string,
    weight:PropTypes.number.isRequired,
    mark:PropTypes.number.isRequired,
  }))),
  dispatch: PropTypes.func,
}

export default Average;
