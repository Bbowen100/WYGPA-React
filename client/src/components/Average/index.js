import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { updateAverage } from '../../actions';
import { Typography } from '@material-ui/core';
import './style.css';

class Average extends Component {
  constructor(props) {
    super(props);
    this.state = {
      average: null
    };
  }

  shouldComponentUpdate(nextProps, nextState) {
    return !(nextState.average === this.state.average);
  }

  calcAvg(source) {
    let avg = 0;
    if (source.items && source.items.length > 0) {
      let totMark = source.items
        .map(ele => ele.mark * ele.weight)
        .reduce((acc, val) => acc + val);
      let totWeight = source.items
        .map(ele => ele.weight)
        .reduce((acc, val) => acc + val);
      avg = Number((totMark / totWeight).toFixed(2));
    }
    if (this.state.average !== avg) {
      this.setState({ average: avg }, () => {
        this.props.dispatch(updateAverage(avg));
      });
    }
  }

  componentDidMount() {
    this.calcAvg(this.props);
  }

  componentWillReceiveProps(newProps) {
    this.calcAvg(newProps);
  }

  render() {
    return (
      <div className="body">
        <Typography variant="headline">Your Average Is</Typography>
        <br />
        <Typography variant="display3">{this.state.average}</Typography>
      </div>
    );
  }
}

Average.propTypes = {
  items: PropTypes.arrayOf(PropTypes.object),
  dispatch: PropTypes.func
};

export default Average;
