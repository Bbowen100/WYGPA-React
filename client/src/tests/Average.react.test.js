import React from 'react';
import Average from '../components/Average';
import renderer from 'react-test-renderer';

test('Average calculates and displays the right average', () => {
  const mockCourses = [
    { name: 'comp1', mark: 78, weight: 1 },
    { name: 'comp2', mark: 77, weight: 2 },
    { name: 'comp3', mark: 76, weight: 1 },
    { name: 'comp4', mark: 75, weight: 2 },
    { name: 'comp5', mark: 74, weight: 1 }
  ];
  const component = renderer.create(
    <Average
      items={mockCourses}
      dispatch={param => {
        console.log(param);
      }}
    />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  //expect(tree.state.average).toEqual(76);
});
