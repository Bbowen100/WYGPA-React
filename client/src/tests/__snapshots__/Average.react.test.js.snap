// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Average calculates and displays the right average 1`] = `
<div
  className="body"
>
  <h1
    className="MuiTypography-root-1 MuiTypography-headline-6"
  >
    Your Average Is
  </h1>
  <br />
  <h1
    className="MuiTypography-root-1 MuiTypography-display3-3"
  >
    76
  </h1>
</div>
`;
